overlord/devicestate/handlers_install.go

```diff
@@ -1214,14 +1251,100 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
 	if err := t.Get("on-volumes", &onVolumes); err != nil {
 		return err
 	}
-	logger.Debugf("install-finish for %q on %v", systemLabel, onVolumes)
-	// TODO: use the seed to get gadget/kernel
-	// - install missing volumes structure content
-	// - install gadget assets
-	// - install kenrel
-	// - make system bootable (include writing modeenv)
+	logger.Debugf("starting install-finish for %q systemd on %v", systemLabel, onVolumes)
+
+	essentialTypes := []snap.Type{snap.TypeKernel, snap.TypeBase, snap.TypeGadget}
+	st.Unlock()
+	sys, snapInfos, snapSeeds, err := m.loadSystemAndEssentialSnaps(systemLabel, essentialTypes)
+	st.Lock()
+	if err != nil {
+		return err
+	}
+	// Mount gadget and kernel
+	mntPtForType := make(map[snap.Type]string)
+	for _, seedSn := range []*seed.Snap{snapSeeds[snap.TypeGadget], snapSeeds[snap.TypeKernel]} {
+		mntPt, rest, err := mountSeedSnap(seedSn)
+		mntPtForType[seedSn.EssentialType] = mntPt
+		if err != nil {
+			return err
+		}
+		defer func() {
+			errRest := rest()
+			if err == nil {
+				err = errRest
+			}
+		}()
+	}
 
-	return fmt.Errorf("finish install step not implemented yet")
+	// Unset revision means the snap will be local, we do something
+	// similar in the seedwriter.
+	// XXX: is this the right place?
+	for _, typ := range essentialTypes {
+		snInfo := snapInfos[typ]
+		if snInfo.Revision.Unset() {
+			snInfo.Revision = snap.R(-1)
+		}
+	}
+
+	// TODO useEncryption always false until setup encryption is implemented
```

Comment:
we want a TODO here that mentions that we probably want to pick/pass a different location for the asset cache are possibly installing something unrelated to the running system, we'll need that passed also into MakeRunnableSystem when we fix the issues in that

Reply:
Added